File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        s1 = []
        s2 = []
        
        while l1 is not None or l2 is not None:
            if l1 is not None:
                s1.append(l1.val)
                l1 = l1.next
            
            if l2 is not None:
                s2.append(l2.val)
                l2 = l2.next
        
        head = None
        carry = 0
        
        while s1 or s2 or carry > 0:
            node = ListNode(carry)
            
            if s1:
                node.val += s1.pop()
            
            if s2:
                node.val += s2.pop()
            
            carry = node.val // 10
            node.val = node.val % 10
            
            node.next = head
            head = node
        
        return head
        

File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /*
        7 2 4 3
        5 6 4
        
        3 4 7 0
        4 6 0 1
        
        curr
        
        0 7
    */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Stack<Integer> s1 = new Stack<Integer>();
        Stack<Integer> s2 = new Stack<Integer>();
        
        while(l1 != null || l2 != null){
            if(l1 != null) {
                s1.push(l1.val);
                l1 = l1.next;
            }
            
            if(l2 != null){
                s2.push(l2.val);
                l2 = l2.next;
            }
        }
        
        ListNode head = null;
        int carry = 0;
        
        while(!s1.isEmpty() || !s2.isEmpty() || carry > 0){
            ListNode node = new ListNode(carry);
            
            if(!s1.isEmpty()) node.val += s1.pop();
            if(!s2.isEmpty()) node.val += s2.pop();
            
            carry = node.val / 10;
            node.val = node.val % 10;
            
            node.next = head;
            head = node;
        }
        
        return head;
    }
}